src/values/acceleration-structure.cxx

```diff
@@ -0,0 +1,2224 @@
+            }
+
+            // Get the children
+            std::vector<std::shared_ptr<Node>> children;
+            for (const auto& child_index : children_indices) {
+                assert(nodes[num_nodes - 1 - child_index] != nullptr);
+                children.push_back(std::move(nodes[num_nodes - 1 - child_index]));
+            }
+
+            nodes.push_back(std::make_shared<BoxNode>(min_bounds, max_bounds, children));
+        }
+
+        // Assertion to make sure nodes list is within expectation
+        unsigned root_index = num_nodes - 1;
+        for (unsigned i = 0; i < nodes.size(); ++i)
+            assert((i != root_index && nodes[i] == nullptr) || (i == root_index && nodes[i] != nullptr));
+
+        // Set the root node
+        root = std::move(nodes[num_nodes - 1]);
+
+        // All shared_ptr in "nodes" should be null, so the root should be null
+        assert(nodes[root_index] == nullptr);
+    };
+
+    AccelerationStructure(const AccelerationStructure& other): id(other.id), isTLAS(other.isTLAS) {
+        activeTrace = other.activeTrace;
+        candidateIntersection = other.candidateIntersection;
+        committedIntersection = other.committedIntersection;
+
+        cullMask = other.cullMask;
+        didPopNodePreviously = other.didPopNodePreviously;
+        nodesToEval = other.nodesToEval;
+        root = other.root->clone();
+
+        useSBT = other.useSBT;
+        missIndex = other.missIndex;
+        offsetSBT = other.offsetSBT;
+        strideSBT = other.strideSBT;
+
+        rayDirection = other.rayDirection;
+        rayOrigin = other.rayOrigin;
+        rayTMin = other.rayTMin;
+        rayTMax = other.rayTMax;
+        rayFlags = other.rayFlags;
+        setFlags();
+    }
+
+private:
+    using ConstNodeRef = const std::shared_ptr<Node>*;
+
+    std::stack<ConstNodeRef> nodesToEval;
+    bool activeTrace = false;
+
+    // Intersection related
+    enum class CommittedIntersectionType { None = 0, Triangle = 1, Generated = 2 };
+    enum class CandidateIntersectionType { Triangle = 0, AABB = 1 };
+
+    struct IntersectionProperties {
+        std::shared_ptr<InstanceNode> instance = nullptr;  // Instance the intersection occured in
+        int geometryIndex = -1;
+        int primitiveIndex = -1;
+        float hitT = std::numeric_limits<float>::max();
+        glm::vec2 barycentrics = glm::vec2(0.0f, 0.0f);
+        bool isOpaque = true;
+        bool enteredTriangleFrontFace = false;
+
+        /// @brief Resets the properties to its default values.
+        void reset() {
+            instance = nullptr;
+            geometryIndex = -1;
+            primitiveIndex = -1;
+            hitT = std::numeric_limits<float>::max();
+            barycentrics = glm::vec2(0.0f, 0.0f);
+            isOpaque = true;
+            enteredTriangleFrontFace = false;
+        }
+
+        IntersectionProperties& operator=(const IntersectionProperties& other) {
```

Comment:
Unless I am mistaken, this method, `CandidateIntersection::operator=`, and `CommittedIntersection::operator=` match the default-generated `operator=` exactly. I would suggest omitting these redundant definitions.